README.md
node-statsd-client
==================

Node.js client for [statsd](https://github.com/etsy/statsd).

[![Build Status](https://secure.travis-ci.org/msiebuhr/node-statsd-client.png?branch=master)](http://travis-ci.org/msiebuhr/node-statsd-client)

Quick tour
----------

```javascript
var SDC = require('statsd-client'),
	sdc = new SDC({host: 'statsd.example.com'});

var timer = new Date();
sdc.increment('some.counter'); // Increment by one.
sdc.gauge('some.gauge', 10); // Set gauge to 10
sdc.timing('some.timer', timer); // Calculates time diff
sdc.histogram('some.histogram', 10, {foo: 'bar'}) // Histogram with tags
sdc.distribution('some.distribution', 10, {foo: 'bar'}) // Distribution with tags

sdc.close(); // Optional - stop NOW
```

API
---

### Initialization

```javascript
var SDC = require('statsd-client'),
	sdc = new SDC({host: 'statsd.example.com', port: 8124});
```

Global options:
 * `prefix`: Prefix all stats with this value (default `""`).
 * `tcp`: User specifically wants to use tcp (default `false`).
 * `socketTimeout`: Dual-use timer. Will flush metrics every interval. For UDP,
   it auto-closes the socket after this long without activity (default 1000 ms;
   0 disables this). For TCP, it auto-closes the socket after `socketTimeoutsToClose` number of timeouts have elapsed without activity.
 * `tags`: Object of string key/value pairs which will be appended on to all StatsD payloads (excluding raw payloads) (default `{}`)

UDP options:
 * `host`: Where to send the stats (default `localhost`).
 * `port`: Port to contact the statsd-daemon on (default `8125`).
 * `ipv6`: Use IPv6 instead of IPv4 (default `false`).

TCP options:
 * `host`: Where to send the stats (default `localhost`).
 * `port`: Port to contact the statsd-daemon on (default `8125`).
 * `socketTimeoutsToClose`: Number of timeouts in which the socket auto-closes if it has been inactive. (default `10`; `1` to auto-close after a single timeout).

HTTP options:
 * `host`: The URL to send metrics to (default: `http://localhost`).
 * `headers`: Additional headers to send (default `{}`)
 * `method`: What HTTP method to use (default `PUT`)

To debug, set the environment variable `NODE_DEBUG=statsd-client` when running your program.

### Counting stuff

Counters are supported, both as raw `.counter(metric, delta)` and with the
shortcuts `.increment(metric, [delta=1])` and `.decrement(metric, [delta=-1])`:

```javascript
sdc.increment('systemname.subsystem.value'); // Increment by one
sdc.decrement('systemname.subsystem.value', -10); // Decrement by 10
sdc.counter('systemname.subsystem.value', 100); // Increment by 100
```

### Gauges

Sends an arbitrary number to the back-end:

```javascript
sdc.gauge('what.you.gauge', 100);
sdc.gaugeDelta('what.you.gauge', 20);  // Will now count 120
sdc.gaugeDelta('what.you.gauge', -70); // Will now count 50
sdc.gauge('what.you.gauge', 10);       // Will now count 10
```

### Sets

Send unique occurences of events between flushes to the back-end:

```javascript
sdc.set('your.set', 200);
```

### Delays

Keep track of how fast (or slow) your stuff is:

```javascript
var start = new Date();
setTimeout(function () {
	sdc.timing('random.timeout', start);
}, 100 * Math.random());
```

If it is given a `Date`, it will calculate the difference, and anything else
will be passed straight through.

And don't let the name (or nifty interface) fool you - it can measure any kind
of number, where you want to see the distribution (content lengths, list items,
query sizes, ...)

Some implementations may require the rate to be explicitly sent for `timing`
(and `histograms`). By default no rate is set, but custom values can be sent.
It is the user's responsibility to control the actual rate.

```javascript
var start = new Date();
sdc.timing('random.timeout', start, { tag: 'value' }, '@0.1');
```

### Histogram

Many implementations (though not the official one from Etsy) support
histograms as an alias/alternative for timers. So aside from the fancy bits
with handling dates, this is much the same as `.timing()`.

### Distribution

Datadog's specific implementation supports another alternative to timers/histograms,
called the [distribution metric type](https://docs.datadoghq.com/metrics/distributions/).
From the client's perspective, this is pretty much an alias to histograms and can be used via `.distribution()`.

### Raw

Passes a raw string to the underlying socket. Useful for dealing with custom
statsd-extensions in a pinch.

```javascript
sdc.raw('foo.bar:123|t|@0.5|#key:value');
```

### Tags

All the methods above support metric level tags as their last argument. Just like global tags, the format for metric tags is an object of string key/value pairs.
Tags at the metric level overwrite global tags with the same key.

```javascript
sdc.gauge('gauge.with.tags', 100, {foo: 'bar'});
```

### Express helper

There's also a helper for measuring stuff in [Express.js](http://expressjs.com)
via middleware:

```javascript
var SDC = require('statsd-client');
var app = express();
	sdc = new SDC({...});

app.use(sdc.helpers.getExpressMiddleware('somePrefix'));
// or
app.get('/',
	sdc.helpers.getExpressMiddleware('otherPrefix'),
	function (req, res, next) { req.pipe(res); });

app.listen(3000);
```

This will count responses by status-code (`prefix.<statuscode>`) and the
overall response-times.

It can also measure per-URL (e.g. PUT to `/:user/:thing` will become
`PUT_user_thing` by setting the `timeByUrl: true` in the `options`-object:

```javascript
app.use(sdc.helpers.getExpressMiddleware('prefix', { timeByUrl: true }));
```

As the names can become rather odd in corner-cases (esp. regexes and non-REST
interfaces), you can specify another value by setting `res.locals.statsdUrlKey`
at a later point.

The `/` page will appear as `root` (e.g. `GET_root`) in metrics while any not found route will appear as `{METHOD}_unknown_express_route`. You can change that name by setting the `notFoundRouteName` in the middleware options.


### Callback helper

There's also a helper for measuring stuff with regards to a callback:

```javascript
var SDC = requrire('statsd-client');
	sdc = new SDC({...});

function doSomethingAsync(arg, callback) {
	callback = sdc.helpers.wrapCallback('somePrefix', callback);
	// ... do something ...
	return callback(null);
}
```

The callback is overwritten with a shimmed version that counts the
number of errors (`prefix.err`) and successes (`prefix.ok`) and
the time of execution of the function (`prefix.time`).
You invoked the shimmed callback exactly the same way as though
there was no shim at all. Yes, you get metrics for your function in
a single line of code.

Note that the start of execution time is marked as soon as you
invoke `sdc.helpers.wrapCallback()`.

You can also provide more options:

```javascript
sdc.helpers.wrapCallback('somePrefix', callback, {
	tags: { foo: 'bar' }
});
```

### Stopping gracefully

By default, the socket is closed if it hasn't been used for a second (see
`socketTimeout` in the init-options), but it can also be force-closed with
`.close()`:

```javascript
var start = new Date();
setTimeout(function () {
	sdc.timing('random.timeout', start); // 2 - implicitly re-creates socket.
	sdc.close(); // 3 - Closes socket after last use.
}, 100 * Math.random());
sdc.close(); // 1 - Closes socket early.
```

The call is idempotent, so you can call it "just to be sure". And if you submit
new metrics later, the socket will automatically be re-created, and a new
timeout-timer started.

### Prefix-magic

The library supports getting "child" clients with extra prefixes, to help with
making sane name-spacing in apps:

```javascript
// Create generic client
var sdc = new StatsDClient({host: 'statsd.example.com', prefix: 'systemname'});
sdc.increment('foo'); // Increments 'systemname.foo'
... do great stuff ...

// Subsystem A
var sdcA = sdc.getChildClient('a');
sdcA.increment('foo'); // Increments 'systemname.a.foo'

// Subsystem B
var sdcB = sdc.getChildClient('b');
sdcB.increment('foo'); // Increments 'systemname.b.foo'
```

Internally, they all use the same socket, so calling `.close()` on any of them
will allow the entire program to stop gracefully.

What's broken
-------------

Check the [GitHub issues](https://github.com/msiebuhr/node-statsd-client/issues).

Other resources
---------------

 * [statsd-tail](https://github.com/msiebuhr/statsd-tail) - A simple program to grab statsd-data on localhost
 * [hot-shots](https://www.npmjs.com/package/hot-shots) - Another popular statsd client for Node.js
 * [statsd](https://github.com/etsy/statsd) - The canonical server

RELEASES
--------

See the [changelog](CHANGELOG.md).

LICENSE
-------

ISC - see
[LICENSE](https://github.com/msiebuhr/node-statsd-client/blob/master/LICENSE).


CHANGELOG.md
### v0.4.7 (2021-06-03)

- [#94](https://github.com/msiebuhr/node-statsd-client/pull/94) feat\(distribution\): add support for datadog's distribution metric ([JGAntunes](mailto:joao@netlify.com))
- [#93](https://github.com/msiebuhr/node-statsd-client/pull/93) Bump hosted-git-info from 2.8.5 to 2.8.9 ([dependabot[bot]](mailto:49699333+dependabot[bot]@users.noreply.github.com))
- [#92](https://github.com/msiebuhr/node-statsd-client/pull/92) Bump lodash from 4.17.19 to 4.17.21 ([dependabot[bot]](mailto:49699333+dependabot[bot]@users.noreply.github.com))

### v0.4.6 (2021-04-22)

#### Pull requests

- [#91](https://github.com/msiebuhr/node-statsd-client/pull/91) update use of Buffer ([khuang](mailto:khuang@havenlife.com))
- [#87](https://github.com/msiebuhr/node-statsd-client/pull/87) Bump lodash from 4.17.11 to 4.17.19 ([dependabot[bot]](mailto:49699333+dependabot[bot]@users.noreply.github.com))
- [#85](https://github.com/msiebuhr/node-statsd-client/pull/85) Add automatically generated changelog ([Andreas Lind](mailto:andreas.lind@peakon.com))

#### Commits to master

- [TCPSocket: Fix spelling \(\#86\)](https://github.com/msiebuhr/node-statsd-client/commit/b230cec5d688c7b84feca4760013be495e3f84e6) ([Morten Siebuhr](mailto:sbhr@sbhr.dk))

### v0.4.4 (2019-09-26)

- [Package: Don't include test\/infrastructure files in NPM pakcage](https://github.com/msiebuhr/node-statsd-client/commit/68e8b91a57fdd5275df0a11e835dfbdb69257dd3) ([Morten Siebuhr](mailto:sbhr@sbhr.dk))

### v0.4.3 (2019-09-26)

#### Pull requests

- [#80](https://github.com/msiebuhr/node-statsd-client/pull/80) Add support for sub routers with parameterized routes ([Garret Meier](mailto:garret@mixmax.com))

#### Commits to master

- [Tests: Skip flappy timing-related test](https://github.com/msiebuhr/node-statsd-client/commit/3f77b0ff931870a957e4556b13b543a1e0356581) ([Morten Siebuhr](mailto:sbhr@sbhr.dk))
- [Travis: Don't test on Node 6.x \(EOL 2019-04-30\)](https://github.com/msiebuhr/node-statsd-client/commit/0edfa2347fdbb68f9b7fd00104eb1398aa3bc7db) ([Morten Siebuhr](mailto:sbhr@sbhr.dk))
- [Travis: Add Node 10 and 12](https://github.com/msiebuhr/node-statsd-client/commit/ac87bf989016423287b4e7b1b52e70199ed0b418) ([Morten Siebuhr](mailto:sbhr@sbhr.dk))
- [Packages: Normailze syntax && add package-lock](https://github.com/msiebuhr/node-statsd-client/commit/f5324f5141d3da8eb65604cb34d7d25567437702) ([Morten Siebuhr](mailto:sbhr@sbhr.dk))
- [Switch to eslint](https://github.com/msiebuhr/node-statsd-client/commit/a15da5ed6f8e359106b70c009853eb98eca9eee8) ([Morten Siebuhr](mailto:sbhr@sbhr.dk))

### v0.4.2
- [Release v0.4.2](https://github.com/msiebuhr/node-statsd-client/commit/5ad537571d768427c61bc4339e0f2c7f4c5d0f32) ([Morten Siebuhr](mailto:sbhr@sbhr.dk))
- [Allow IPv6 UDP](https://github.com/msiebuhr/node-statsd-client/commit/f0815a0f4707a4ba4470b8c1fb5a828972fd8630) ([Ceesjan Luiten](mailto:ceesjan@ytec.nl))

### v0.4.1
- [Release v0.4.1](https://github.com/msiebuhr/node-statsd-client/commit/c21843a7bc8d7b67711d4a9864047dcece2232e0) ([Morten Siebuhr](mailto:sbhr@sbhr.dk))
- [README: Add some external links](https://github.com/msiebuhr/node-statsd-client/commit/f7d887573f75f1866af07bfa011064e7192946de) ([Morten Siebuhr](mailto:sbhr@sbhr.dk))
- [Don't test on 0.12 any more...](https://github.com/msiebuhr/node-statsd-client/commit/419330fc4c95c298a0e904539eaaa70b24e88c5c) ([Morten Siebuhr](mailto:sbhr@sbhr.dk))
- [Test on Node 4, 6, 8 and latest](https://github.com/msiebuhr/node-statsd-client/commit/3df7eb439a36e1201f1ad1e272491660a5e265a0) ([Morten Siebuhr](mailto:sbhr@sbhr.dk))
- [Add callback helper](https://github.com/msiebuhr/node-statsd-client/commit/b828eb675a3c3968c88b44f0741b74637039b16d) ([GochoMugo](mailto:mugo@forfuture.co.ke))

### v0.4.0
- [Release v0.4.0](https://github.com/msiebuhr/node-statsd-client/commit/98ea30571806e68a0badcdaeecfef2790977be08) ([Morten Siebuhr](mailto:sbhr@sbhr.dk))
- [Merge tags with 0.12 valid functions](https://github.com/msiebuhr/node-statsd-client/commit/a0c526a3dc3b0760401a95a34c98c73a69de4fbf) ([Chris911](mailto:christophe.naud.dulude@gmail.com))
- [Update README for metric level tags](https://github.com/msiebuhr/node-statsd-client/commit/1e643efa6f2b0d526014e8c0d78e9dd2e0254d20) ([Chris911](mailto:christophe.naud.dulude@gmail.com))
- [Add metrics level tags](https://github.com/msiebuhr/node-statsd-client/commit/1bf097b0e1367dce08ca0877e0f2a8ec7508ec61) ([Chris911](mailto:christophe.naud.dulude@gmail.com))
- [Add .jshintignore](https://github.com/msiebuhr/node-statsd-client/commit/f3609d699b3b92ae88ec6332eaf6a67dc71eb1df) ([Andreas Lind](mailto:andreas@one.com))
- [+3 more](https://github.com/msiebuhr/node-statsd-client/compare/v0.3.0...v0.4.0)

### v0.3.0
#### Pull requests

- [#68](https://github.com/msiebuhr/node-statsd-client/pull/68) Fixed typo in TCPSocket for \_maxBufferSize property ([Philippe Guilbault](mailto:philippe.guilbault@playster.com))

#### Commits to master

- [Release v0.3.0](https://github.com/msiebuhr/node-statsd-client/commit/260f56ee29b83d39036462a87d6efb7d113fce73) ([Morten Siebuhr](mailto:sbhr@sbhr.dk))
- [Add support for StatsD tags on all payloads \(\#69\)](https://github.com/msiebuhr/node-statsd-client/commit/9e7756eb046a4fa9cb2a058328bfd3ab63a5bafb) ([Brian Conn](mailto:bcconn2112@gmail.com))
- [Enable express middleware to send both url prefixed and non prefixed status code increments \(\#67\)](https://github.com/msiebuhr/node-statsd-client/commit/9cfe31b2f14ddd12016e8ff97633171e00f170d6) ([Matthew Sessions](mailto:shichongrui@gmail.com))

### v0.2.4
#### Pull requests

- [#66](https://github.com/msiebuhr/node-statsd-client/pull/66) Replace anonymous functions with named functions ([Oleksandr Kureniov](mailto:s@zxc.pp.ua))

#### Commits to master

- [Release v0.2.4](https://github.com/msiebuhr/node-statsd-client/commit/0aa554430ae2ced8a6f55550c61682007d076013) ([Morten Siebuhr](mailto:sbhr@sbhr.dk))

### v0.2.3
#### Pull requests


lib/TCPSocket.js
var net = require('net');
var debug = require('util').debuglog('statsd-client');

/*global console*/

function TCPSocket(options) {
    options = options || {};

    this._hostname = options.host || 'localhost';
    this._port = options.port || 8125;
    this._flushBufferTimeout = 'socketTimeout' in options ? options.socketTimeout : 1000;
    this._timeoutsToClose = 'socketTimeoutsToClose' in options ? options.socketTimeoutsToClose : 10; // close socket if not used in 10 flush intervals

    // Check https://github.com/etsy/statsd/#multi-metric-packets for advisable sizes.
    this._maxBufferSize = 'maxBufferSize' in options ? options.maxBufferSize : 1200;

    // Set up re-usable socket
    this._socket = undefined; // Store the socket here
    this._socketUsed = false; // Flag if it has been used
    this._socketLastUsed = 0; // How many intervals of timeout since socket has been used
    this._socketTimer = undefined; // Reference to check-timer
    this._buffer = [];
}

/* Dual-use timer.
 *
 * First checks if there is anything in it's buffer that need to be sent. If it
 * is non-empty, it will be flushed. (And thusly, the socket is in use and we
 * stop checking further right away).
 *
 * If there is nothing in the buffer and the socket hasn't been used in the
 * previous interval, close it.
 */
TCPSocket.prototype._socketTimeout = function _socketTimeout() {
    debug("close()");
    // Flush the buffer, if it contain anything.
    if (this._buffer.length > 0) {
        this._flushBuffer();
        return;
    }

    // Is it already closed? -- then stop here
    if (!this._socket) {
        return;
    }

    // Not used?
    if (this._socketUsed === false) {
        this._socketLastUsed++;
        // if not used in many intervals, close it
        if (this._socketLastUsed >= this._timeoutsToClose) {
            this.close();
            return;
        }
    } else {
        this._socketLastUsed = 0;
    }

    // Reset whether its been used
    this._socketUsed = false;
    // Start timer, if we have a positive timeout
    if (this._flushBufferTimeout > 0 && !this._socketTimer) {
        this._socketTimer = setInterval(this._socketTimeout.bind(this), this._flushBufferTimeout);
    }
};


/*
 * Close the socket, if in use and cancel the interval-check, if running.
 */
TCPSocket.prototype.close = function close() {
    debug("close()");
    if (!this._socket) {
        return;
    }

    if (this._buffer.length > 0) {
        this._flushBuffer();
    }

    // Cancel the running timer
    if (this._socketTimer) {
        clearInterval(this._socketTimer);
        this._socketTimer = undefined;
    }

    // Wait a tick or two, so any remaining stats can be sent.
    setTimeout(this.kill.bind(this), 10);
};

/* Kill the socket RIGHT NOW.
 */
TCPSocket.prototype.kill = function kill() {
    debug("kill()");
    if (!this._socket) {
        return;
    }

    // Clear the timer and catch any further errors silently
    if (this._socketTimer) {
        clearInterval(this._socketTimer);
        this._socketTimer = undefined;
    }
    this._socket.on('error', function () {});

    this._socket.end();
    this._socket = undefined;
};

TCPSocket.prototype._createSocket = function _createSocket(callback) {
    debug("_createSocket()");
    if (this._socket) {
        return callback();
    }

    this._socket = net.Socket({
        type: 'tcp4'
    });


    // Listen on 'error'-events, so they don't bubble up to the main
    // application. Try closing the socket for now, forcing it to be re-created
    // later.
    this._socket.once('error', this.kill.bind(this));

    // Call on when the socket is ready.
    this._socket.connect(this._port, this._hostname, function() {
        return callback();
    });

    // Start timer, if we have a positive timeout
    if (this._flushBufferTimeout > 0 && !this._socketTimer) {
        this._socketTimer = setInterval(this._socketTimeout.bind(this), this._flushBufferTimeout);
    }
};

/* Buffer management
 */
TCPSocket.prototype._enqueue = function _enqueue(data) {
    debug("_enqueue(", data, ")");

    if (!this._socketTimer) {
        this._socketTimer = setInterval(this._socketTimeout.bind(this), this._flushBufferTimeout);
    }
    // Empty buffer if it's too full
    if (this._buffer.reduce(function(sum, line) { return sum + line.length; }, 0) > this._maxBufferSize) {
        this._flushBuffer();
    }

    this._buffer.push(data);
};

TCPSocket.prototype._flushBuffer = function _flushBuffer() {
    debug("_flushBuffer() →", this._buffer);
    var that = this;
    this._send(this._buffer, function() {
        that._buffer = [];
    });
};

/* Send data - public interface.
 */
TCPSocket.prototype.send = function send(data) {
    debug("send(", data, ")");
    if (this._maxBufferSize === 0) {
        return this._send([data], function() {});
    } else {
        this._enqueue(data);
    }
};

/*
 * Send data.
 */
TCPSocket.prototype._send = function _send(data, callback) {
    debug("_send(", data, ")");
    // If we don't have a socket, or we have created one but it isn't
    // ready yet, we need to enqueue data to send once the socket is ready.
    var that = this;

    //Must be asynchronous
    this._createSocket(that._writeToSocket.bind(that, data, callback));
};

TCPSocket.prototype._writeToSocket = function _writeToSocket(data, callback) {
    // Create message
    // Trailing \n important because socket.write will sometimes concat multiple 'write' calls.
    var message = new Buffer(data.join('\n') + '\n');

    debug(message.toString());
    if (this._socket) {
        this._socketUsed = true;
        this._socket.write(message);
    }

    callback();
};

module.exports = TCPSocket;
